tmt/lint.py

```diff
@@ -0,0 +1,323 @@
+            linters = [
+                linter for linter in linters
+                if linter.id not in disable_checks
+                ]
+
+        return linters
+
+    def lint(
+            self,
+            enable_checks: Optional[List[str]] = None,
+            disable_checks: Optional[List[str]] = None,
+            enforce_checks: Optional[List[str]] = None) -> Tuple[bool, List[LinterRuling]]:
+        """
+        Check the instance against a battery of linters and report results.
+
+        :returns: whether the instance is valid.
+        """
+
+        enforce_checks = enforce_checks or []
+
+        linters = self.resolve_enabled_linters(
+            enable_checks=enable_checks,
+            disable_checks=disable_checks)
+
+        valid = True
+        rulings: List[LinterRuling] = []
+
+        for linter in sorted(linters, key=lambda x: x.id):
+            for outcome, message in linter.callback(self):
+                if outcome == LinterOutcome.FAIL:
+                    rulings.append((linter, outcome, outcome, message))
+
+                    valid = False
+
+                elif outcome != LinterOutcome.PASS:
+                    if linter.id in enforce_checks:
+                        rulings.append((linter, outcome, LinterOutcome.FAIL, message))
+
+                        valid = False
+
+                    else:
+                        rulings.append((linter, outcome, outcome, message))
+
+                else:
+                    rulings.append((linter, outcome, outcome, message))
+
+        return valid, rulings
+
+    @classmethod
+    def format_linters(cls) -> str:
+        hints: List[str] = []
+
+        for linter in sorted(cls.get_linter_registry(), key=lambda x: x.id):
+            # This has been already verified when registering the linter
+            assert linter.callback.__doc__ is not None
+
+            hints.append(textwrap.dedent(linter.callback.__doc__).strip())
+
+        return '\n'.join(hints)
+
+
+def filter_allowed_checks(
+        rulings: Iterable[LinterRuling],
+        outcomes: Optional[List[LinterOutcome]] = None) -> Generator[LinterRuling, None, None]:
+    """
+    Filter only rulings whose outcomes are allowed.
+
+    :param rulings: rulings to process.
+    :param outcomes: a list of allowed ruling outcomes, :py:class:`LinterOutcome`.
+        If not set, all outcomes are allowed.
+    :yields: rulings with allowed outcomes.
+    """
+
+    outcomes = outcomes or []
+
+    for linter, actual_outcome, eventual_outcome, message in rulings:
+        if outcomes and eventual_outcome not in outcomes:
+            continue
+
+        yield (linter, actual_outcome, eventual_outcome, message)
+
+
+def format_rulings(rulings: Iterable[LinterRuling]) -> Generator[str, None, None]:
+    """
+    Format rulings for printing or logging.
+
+    :param rulings: rulings to format.
+    :yields: rulings formatted as separate strings.
+    """
+
+    display_eventual = any(
```

Comment:
Perhaps a short comment here explaining what `display_eventual` is would be handy.

Reply:
Added in 21c4e01c.